python/metatensor-torch/metatensor/torch/atomistic/ase_calculator.py

```diff
@@ -0,0 +1,322 @@
+        # Compute the neighbors lists requested by the model using ASE NL
+        for options in self._model.requested_neighbors_lists():
+            system.add_neighbors_list(
+                options, neighbors=_compute_ase_neighbors(atoms, options)
+            )
+
+        return self._model(
+            [system],
+            run_options,
+            check_consistency=self.parameters["check_consistency"],
+        )
+
+    def calculate(
+        self,
+        atoms: ase.Atoms,
+        properties: List[str],
+        system_changes: List[str],
+    ) -> Dict[str, np.ndarray]:
+        """
+        Compute some ``properties`` with this calculator, and return them in the format
+        expected by ASE.
+
+        This is not intended to be called directly by users, but to be an implementation
+        detail of ``atoms.get_energy()`` and related functions. See
+        :py:meth:`ase.calculators.calculator.Calculator.calculate` for more information.
+        """
+        super().calculate(
+            atoms=atoms,
+            properties=properties,
+            system_changes=system_changes,
+        )
+
+        outputs = _ase_properties_to_metatensor_outputs(properties)
+        species, positions, cell = _ase_to_torch_data(atoms)
+
+        do_backward = False
+        if "forces" in properties:
+            do_backward = True
+            positions.requires_grad_(True)
+
+        if "stress" in properties:
+            do_backward = True
+
+            scaling = torch.eye(3, requires_grad=True, dtype=cell.dtype)
+
+            positions = positions @ scaling
+            positions.retain_grad()
+
+            cell = cell @ scaling
+
+        if "stresses" in properties:
+            raise NotImplementedError("'stresses' are not implemented yet")
+
+        # convert from ase.Atoms to metatensor.torch.atomistic.System
+        system = System(species, positions, cell)
+        for options in self._model.requested_neighbors_lists():
+            system.add_neighbors_list(
+                options, neighbors=_compute_ase_neighbors(atoms, options)
+            )
+
+        run_options = ModelEvaluationOptions(
+            length_unit="angstrom",
+            outputs=outputs,
+            selected_atoms=None,
+        )
+
+        outputs = self._model(
+            [system],
+            run_options,
+            check_consistency=self.parameters["check_consistency"],
+        )
+        energy = outputs["energy"]
+
+        if run_options.outputs["energy"].per_atom:
+            assert len(energy) == 1
+            assert energy.sample_names == ["structure", "atom"]
+            assert torch.all(energy.block().samples["structure"] == 0)
+            assert torch.all(
+                energy.block().samples["atom"] == torch.arange(positions.shape[0])
+            )
+            energies = energy.block().values
+            assert energies.shape == (len(atoms), 1)
+
+            energy = sum_over_samples(energy, sample_names=["atom"])
```

Comment:
@frostedoyster regarding how to handle having the same output both per-atom and per-structure, this is how it's done in the ASE calculator. Feels to me that this is simple enough to not require more infrastructure than what we have. If the end user wants both, compute per-atom & then sum over samples to get the per-structure.

Reply:
Yes, I feel like most models do the summation internally. In that case, they can interface with us (wrap) and they won't support per-atom energies. So it seems to me that, with the current design, the summation is supposed to be done by the model, while any gradient calls would happen outside. I still think it would be cleaner (in metatensor-models) to have the model output the forces as well though, using the "explicit gradients" that we have here

Reply:
I'm not sure I'm following. If the model does not support per-atom output, it can declare it, and then we will raise an error if the engine request per-atom output.

Same with explicit gradients: if the model supports it, it can declare it and then the engine can use it. Otherwise the engine falls back to backward propagation, which should always be supported.